File: frontend/src/apis/contest/EvaluateContestProblemVisibility.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { ProblemId } from '@/objects/problem/ProblemId'

type EvaluateContestProblemVisibilityBody = {
  problemId: ProblemId
  submittedAt: string | null
}

export class EvaluateContestProblemVisibility implements APIWithSessionMessage<unknown> {
  declare readonly responseType?: unknown
  readonly method = 'POST'
  readonly apiPath = 'internal/contests/evaluate-problem-visibility'
  private readonly problemId: ProblemId
  private readonly submittedAt: string | null

  constructor(problemId: ProblemId, submittedAt: string | null) {
    this.problemId = problemId
    this.submittedAt = submittedAt
  }

  body(): EvaluateContestProblemVisibilityBody {
    return {
      problemId: this.problemId,
      submittedAt: this.submittedAt,
    }
  }
}

File: backend/src/main/scala/domains/contest/api/EvaluateContestProblemVisibility.scala
package domains.contest.api

import cats.effect.IO
import domains.auth.api.InternalOnlyAuthenticatedApi
import domains.auth.objects.internal.AuthenticatedUser
import domains.contest.table.contest.ContestProblemVisibilityTable
import domains.problem.objects.ProblemId
import org.http4s.Method
import shared.api.ApiPath

import java.sql.Connection
import java.time.Instant

object EvaluateContestProblemVisibility
  extends InternalOnlyAuthenticatedApi[EvaluateContestProblemVisibility.Input, EvaluateContestProblemVisibility.Result]:

  final case class Input(
    problemId: ProblemId,
    submittedAt: Option[Instant]
  )

  final case class Result(
    hasVisibleUnfinishedContestContainingProblem: Boolean,
    hasVisibleEndedContestContainingProblem: Boolean,
    hasRegisteredContestContainingSubmission: Boolean
  )

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/internal/contests/evaluate-problem-visibility")

  override def plan(connection: Connection, actor: AuthenticatedUser, input: Input): IO[Result] =
    for
      hasVisibleUnfinishedContestContainingProblem <- ContestProblemVisibilityTable
        .hasVisibleUnfinishedContestContainingProblem(connection, actor, input.problemId)
      hasVisibleEndedContestContainingProblem <- ContestProblemVisibilityTable
        .hasVisibleEndedContestContainingProblem(connection, actor, input.problemId)
      hasRegisteredContestContainingSubmission <- input.submittedAt match
        case Some(submittedAt) =>
          ContestProblemVisibilityTable.hasRegisteredContestContainingSubmission(connection, actor, input.problemId, submittedAt)
        case None =>
          IO.pure(false)
    yield Result(
      hasVisibleUnfinishedContestContainingProblem = hasVisibleUnfinishedContestContainingProblem,
      hasVisibleEndedContestContainingProblem = hasVisibleEndedContestContainingProblem,
      hasRegisteredContestContainingSubmission = hasRegisteredContestContainingSubmission
    )
